### PR TITLE
docs: add git workflow

### PR DESCRIPTION
**Pull request by apahim**
Create a git workflow section based on the devel list discussion:
https://www.redhat.com/archives/avocado-devel/2016-May/msg00057.html

### CODE REVIEW

File: docs/source/ContributionGuide.rst [hunk 1]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+
+- Golden tip: never work on local branch master. Instead, create a new
+  local branch and checkout to it::
+
+    $ git checkout -b my_new_local_branch
+
+- Code and then commit your changes::
+
+    $ git add new-file.py
+    $ git commit -a -s
```
Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: I'd skip the `-a` as you already added the changes by `git add`. (alternatively I'd mention use:

```
$ git add new-file.py
$ git commit -s
# or "git commit -as" to commit all changes
```

File: docs/source/ContributionGuide.rst [hunk 2]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+
+- Write a good commit message, pointing motivation, issues that you're
+  addressing. Usually you should try to explain 3 points in the commit
+  message: motivation, approach and effects::
+
+    header          <- Limited to 50 characters. No period.
```
Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: I thought it's 70

Comment (apahim) [inline docs/source/ContributionGuide.rst]: Github limit is 72. Gitweb is 50.

Comment (apahim) [inline docs/source/ContributionGuide.rst]: (iirc)

File: docs/source/ContributionGuide.rst [hunk 3]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+
+- Push your commit(s) to your fork::
+
+    $ git push origin my_new_local_branch
+
+- Create the Pull Request on github. Add the relevant information to the
+  Pull Request description.
+
+- In the Pull Request discussion page, comment with the link to the
+  ``job.log`` pastebin.
```
Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: output/file instead of pastebin would be probably better

File: docs/source/ContributionGuide.rst [hunk 4]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+- Make sure your code is working (install your version of avocado, test
+  your change, run ``make check`` to make sure you didn't introduce any
+  regressions).
+
+- Paste the ``job.log`` file content from the previous step in a pastebin
+  service, like fpaste.org.
+
+- Push your commit(s) to your fork::
+
+    $ git push origin my_new_local_branch
```
Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: I'd recommend `git fetch && git rebase upstream/master`

File: docs/source/ContributionGuide.rst [hunk 5]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+- Code, and amend the commit(s) and/or create new commits. If you have
+  more than one commit in the PR, you will probably need to rebase
+  interactively to amend the right commits. ``git cola`` or ``git citool``
+  can be handy here.
+
+- Rebase your local branch on top of upstream master::
+
+    $ git checkout master
+    $ git pull upstream master
+    $ git push
```
Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: I never ever needed to do this :smile:. Simple `git fetch && git rebase origin/master` works well :wink:.

File: docs/source/ContributionGuide.rst [hunk 6]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+  Pull Request introduced when compared to the previous Pull Request(s).
+
+- Close the previous Pull Request on github.
+
+After your PR gets merged, you can sync your local repository and your
+fork on github::
+
+    $ git checkout master
+    $ git pull upstream master
+    $ git push
```
Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: I don't see a reason for this part...

File: docs/source/ContributionGuide.rst [hunk 7]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+                    <- Blank line
+    Reference:      <- External references, one per line (issue, trello, ...)
+    Signed-off-by:  <- Signature (created by git commit -s)
+
+- Make sure your code is working (install your version of avocado, test
+  your change, run ``make check`` to make sure you didn't introduce any
+  regressions).
+
+- Paste the ``job.log`` file content from the previous step in a pastebin
+  service, like fpaste.org.
```
Comment (clebergnu) [inline docs/source/ContributionGuide.rst]: One little tip that may be useful to include (in a "hint" like section) is to install/use the `fpaste` utility.  Running `fpaste ~/avocado/job-results/latest/job.log` can be rather convenient.

File: docs/source/ContributionGuide.rst [hunk 8]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+
+From time to time, you can remove old branches to avoid pollution::
+
+    $ git push origin :my_old_branch
+
 Signing commits
 ---------------
 
-Optionally we encourage people to sign the commits using GPG signatures. Doing
+Optionally you can to sign the commits using GPG signatures. Doing
```
Comment (clebergnu) [inline docs/source/ContributionGuide.rst]: "You can sign", that is, `s/to//`.

File: docs/source/ContributionGuide.rst [hunk 9]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+After your PR gets merged, you can sync your local repository and your
+fork on github::
+
+    $ git checkout master
+    $ git pull upstream master
+    $ git push
+
+From time to time, you can remove old branches to avoid pollution::
+
+    $ git push origin :my_old_branch
```
Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: Unlike the previous part I like this section. I'd even add:

```
# To list branches along with time reference
$ git for-each-ref --sort='-authordate:iso8601' --format=' %(authordate:iso8601)%09%(refname)' refs/heads
# To list merged branches
$ git branch --merged
```

PS: This only works if you `push` changes to your `origin`, or if you use the `origin` branch as upstream (which is what I do)

File: docs/source/ContributionGuide.rst (hunk 2 above)
Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: Well 50 is really short...

File: docs/source/ContributionGuide.rst (hunk 1 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: ack

File: docs/source/ContributionGuide.rst (hunk 2 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: Let's keep 72, since github can handle that.

File: docs/source/ContributionGuide.rst (hunk 7 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: ack

File: docs/source/ContributionGuide.rst (hunk 4 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: ack

File: docs/source/ContributionGuide.rst (hunk 3 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: ok

File: docs/source/ContributionGuide.rst (hunk 5 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: Don't you update your fork's master branch? Mine is always in sync :)

File: docs/source/ContributionGuide.rst (hunk 6 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: Sync from upstream to local master and then from local master to origin(fork) master. After that, the 3 master branches are in sync: upstream(project repo), origin(user fork) and local.

File: docs/source/ContributionGuide.rst (hunk 9 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: ok

File: docs/source/ContributionGuide.rst (hunk 8 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: Yep.

File: docs/source/ContributionGuide.rst (hunk 6 above)
Comment (apahim) [inline docs/source/ContributionGuide.rst]: Yes, I see your fork's master branch is 2 years old :)
Anyway, I have the habit to keep my fork's master branch in sync.

Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: Sure, no strong opinion, it's just a workflow example :wink:

File: docs/source/ContributionGuide.rst [hunk 10]
```diff
@@ -65,10 +65,119 @@ be submitted as a github Pull Request. This process tries to ensure that every
...
+
+- In the Pull Request discussion page, comment with the link to the
+  ``job.log`` pastebin.
+
+- Check if your Pull Request passes the CI (travis). Your Pull Request
+  will probably be ignored until it's all green.
+
+Now you're waiting for feedback on github Pull Request page. Once you
+get some, join the discussion, answer the questions, make clear if you're
+going to change the code based on some review and, if not, why. New versions
```
Comment (ldoktor) [inline docs/source/ContributionGuide.rst]: Over the night I thought we might want to explicitly say `Feel free to disagree with the reviewer, they probably have different usecases and opinions, which is expected. Try describing yours and suggest other solutions, if necessary.`

... or something which reminds people that they do not need to fix everything, but in the end both, author and reviewer has to reach a conclusion.

Comment (apahim) [inline docs/source/ContributionGuide.rst]: good one

**Pull request closed without merging**